include/circt/Dialect/Calyx/CalyxLoweringUtils.h

```diff
@@ -84,11 +84,13 @@ void buildAssignmentsForRegisterWrite(OpBuilder &builder,
 // A structure representing a set of ports which act as a memory interface for
 // external memories.
 struct MemoryPortsImpl {
-  Value readData;
-  Value done;
-  Value writeData;
+  std::optional<Value> readData;
+  std::optional<Value> readEn;
+  std::optional<Value> readDone;
+  std::optional<Value> writeData;
+  std::optional<Value> writeEn;
+  std::optional<Value> writeDone;
```

Comment:
What's the reason for introducing all of these `std::optional` wrappers and `get`ter methods, given that a `Value` is nullable?

Reply:
I guess that's a fair point, I just generally prefer being explicit about optionality when possible. The whole point of structuring things like this is to handle different kinds of memory interfaces (read only, different kinds of read/write enable/done, etc.) so these values are explicitly optional at times, whereas I often think of a null Value as an error state. If this is bad style for CIRCT, let me know and I can change.

Reply:
I think this is somewhat a matter of code style, but I personally prefer optional made explicit as it is here. You could use null Values, but if you ever pass a null Value to MLIR APIs you will quickly hit asserts, or segfaults. Making optionality explicit like this lets the type system enforce that you're considering absence at the right places. This isn't totally fair, but the question is kind of like asking why Java introduced Optional given that references are nullable.

Reply:
> LLVM tends to prefer not having redundant checking code to avoid "slowdown by a thousand minor inefficiencies".

Reference?

First, I too am in favor of using explicit Optional instead of depending on nullability; they have two different meanings. Second, I think we should prove this is actually causing a performance regression before talking about inefficiencies. (In general, I agree with Mike - it boils down to preference. There are arguments both for and against.)

Reply:
The worry about redundant code checks is not really relevant in this case anyways, because we will regularly have to check what ports a memory interface has to make decisions about how to connect to a given interface. We're just using optional to check this instead of explicitly checking for a null Value.